CRM/MembershipExtras/Hook/Post/ContributionRecur.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+/**
+ * Implements post-process hooks on ContributionRecur entity.
+ */
+class CRM_MembershipExtras_Hook_Post_ContributionRecur {
+
+  /**
+   * Reference to the recurring contribution's BAO that was stored.
+   *
+   * @var \CRM_Contribute_BAO_ContributionRecur
+   */
+  private $contributionRecurBAO;
+
+  /**
+   * CRM_MembershipExtras_Hook_Post_ContributionRecur constructor.
+   *
+   * @param \CRM_Contribute_BAO_ContributionRecur $contributionBAO
+   */
+  public function __construct(CRM_Contribute_BAO_ContributionRecur $contributionBAO) {
+    $this->contributionRecurBAO = $contributionBAO;
+  }
+
+  /**
+   * Post processes recurring contribution entity.
+   */
+  public function postProcess() {
+    $this->updateLineItemEndDates();
+  }
+
+  /**
+   * Updates end dates for line items associated to the recurring contribution.
+   */
+  private function updateLineItemEndDates() {
+    $contributionStatus = CRM_Contribute_PseudoConstant::contributionStatus($this->contributionRecurBAO->contribution_status_id, 'name');
+
+    if ($contributionStatus === 'Completed' && $this->contributionRecurBAO->installments > 1) {
```

Comment:
is : 

" $this->contributionRecurBAO->installments > 1"

needed ? since even recur contiution where installments = NULL will have one installment which mean and one or more recur line items.

Reply:
Very needed! Line items for recurring contributions with 1 or less installments should NEVER have their end date set when payment is completed.

End dates for line items for payment plans with 1 or less installments are set on renewal... I know, this is BS, but, that's how it goes...